src/client/metric.rs

```diff
@@ -155,37 +229,132 @@ macro_rules! check_dim (
     }
 
     /// Modifies and returns the unit with given space scale and dimension
     pub fn space(mut self, scale: Space, dim: i8) -> Result<Self, String> {
         check_dim!(dim);
-        self.pmapi_repr |= (scale as u32) << 16;
-        self.pmapi_repr |= ((dim as u32) & ((1 << 4) - 1)) << 28;
+        self.pmapi_repr |= (scale as u32) << SPACE_SCALE_LSB;
+        self.pmapi_repr |= ((dim as u32) & LS_FOUR_BIT_MASK) << SPACE_DIM_LSB;
         Ok(self)
     }
 
     /// Modifies and returns the unit with given time scale and dimension
     pub fn time(mut self, time: Time, dim: i8) -> Result<Self, String> {
         check_dim!(dim);
-        self.pmapi_repr |= (time as u32) << 12;
-        self.pmapi_repr |= ((dim as u32) & ((1 << 4) - 1)) << 24;
+        self.pmapi_repr |= (time as u32) << TIME_SCALE_LSB;
+        self.pmapi_repr |= ((dim as u32) & LS_FOUR_BIT_MASK) << TIME_DIM_LSB;
         Ok(self)
     }
 
     /// Modifies and returns the unit with given count scale and dimension
     pub fn count(mut self, count: Count, dim: i8) -> Result<Self, String> {
         check_dim!(dim);
-        self.pmapi_repr |= (count as u32) << 8;
-        self.pmapi_repr |= ((dim as u32) & ((1 << 4) - 1)) << 20;
+        self.pmapi_repr |= (count as u32) << COUNT_SCALE_LSB;
+        self.pmapi_repr |= ((dim as u32) & LS_FOUR_BIT_MASK) << COUNT_DIM_LSB;
         Ok(self)
     }
+
+    fn space_scale(&self) -> u8 {
+        ((self.pmapi_repr >> SPACE_SCALE_LSB) & LS_FOUR_BIT_MASK) as u8
+    }
+
+    fn time_scale(&self) -> u8 {
+        ((self.pmapi_repr >> TIME_SCALE_LSB) & LS_FOUR_BIT_MASK) as u8
+    }
+
+    fn count_scale(&self) -> u8 {
+        ((self.pmapi_repr >> COUNT_SCALE_LSB) & LS_FOUR_BIT_MASK) as u8
+    }
+
+    /*
+        We have a 4-bit value in two's complement form which we have to
+        sign-extend to 8 bits. We first left shift our 4 bits in pmapi_repr
+        to the most significant position, then do an arithmetic right shift
+        to bring them to the least significant position in order to
+        sign-extend the remaining 4 bits.
+
+        In Rust, an arthimetic/logical right shift is performed depending on
+        whether the integer is signed or unsigned. Hence, we cast our integer
+        to an i32 before we right shift it.
+    */
+    fn dim(&self, lsb: u8) -> i8 {
+        (
+            ( self.pmapi_repr << (32 - (lsb + 4)) ) as i32
+            >> 28
+        ) as i8
+    }
+
+    fn space_dim(&self) -> i8 {
+        self.dim(SPACE_DIM_LSB)
+    }
+
+    fn time_dim(&self) -> i8 {
+        self.dim(TIME_DIM_LSB)
+    }
+
+    fn count_dim(&self) -> i8 {
+        self.dim(COUNT_DIM_LSB)
+    }
+}
+
+macro_rules! write_dim (
+    ($dim:expr, $scale:expr, $scale_type:tt, $f:expr) => (
+        if let Some(dim_scale) = $scale_type::from_u8($scale) {
+            write!($f, "{}", dim_scale)?;
+            if $dim > 1 {
+                write!($f, "^{}", $dim.abs())?;
+            }
+            write!($f, " ")?;
```

Comment:
remove this space, this is causing the output in test case 5 to be `MiB  / sec` (extra space after `MiB`)